### PR TITLE
Refactor SOCI index builder code into a separate module.

### PR DESCRIPTION
**Pull request by dvnguyen-amzn**
*Issue #, if available:* #128

*Description of changes:*
Refactor SOCI index builder code into a separate module.
Motivation: to allow reusing the index builder code in other projects (issue #128).

This PR contains minimal change to existing soci_index.go. If this PR is accepted, we'll have another follow up PR that replaces existing index builder code with the newly created module.

*Testing performed:*
- running `make && make test && make check`
- writing an example code and use the module locally to build soci index for a pulled image

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: soci/index_builder.go [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+	"github.com/containerd/containerd/images"
+	"github.com/containerd/containerd/platforms"
+	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
+	"github.com/pkg/errors"
+	"golang.org/x/net/context"
+	"golang.org/x/sync/errgroup"
+)
+
+type IndexBuilder struct {
+	contentStore        content.Store
```
Comment (rdpsin) [inline soci/index_builder.go:18]: Do we want to use `content.Store` here? As @Kern-- mentioned, `content.Store` might be too large of an interface to use here. We can instead use ORAS's [content.Storage](https://pkg.go.dev/oras.land/oras-go/v2@v2.0.0-rc.4/content#Storage) or define our own.

File: soci/index_builder.go [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+	spanSize            int64
+	minLayerSize        int64
+	buildToolIdentifier string
+}
+
+func NewIndexBuilder(contentStore content.Store) *IndexBuilder {
+	defaultSpanSize := int64(1 << 22)
```
Comment (rdpsin) [inline soci/index_builder.go:25]: We should define a const up top for defaultSpanSize

```
const (
    defaultSpanSize = 1 << 22
 )
 ```

**Comment by rdpsin** [comment]
Can you run `make add-ltag` to add headers to new files?

File: soci/index_builder.go [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+	return &IndexBuilder{
+		contentStore:        contentStore,
+		spanSize:            defaultSpanSize,
+		minLayerSize:        0,
+		buildToolIdentifier: "AWS SOCI CLI v0.1",
```
Comment (hanyuel) [inline soci/index_builder.go:30]: We probably should have the const string "AWS SOCI CLI v0.1"d defined else where and have the constructor pass in ```buildToolIdentifier```.

File: soci/index_builder.go [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+}
+
+func (b *IndexBuilder) WithBuildToolIdentifer(buildToolIdentifier string) *IndexBuilder {
+	b.buildToolIdentifier = buildToolIdentifier
+	return b
+}
+
+// BuildIndex builds SOCI index for an image
+// This returns SOCI index manifest and an array of io.Reader of zTOCs
+func (b *IndexBuilder) BuildIndex(ctx context.Context, desc ocispec.Descriptor) (Index, []io.Reader, error) {
```
[the remark this answers is not in the document]
Comment (dvnguyen-amzn) [reply soci/index_builder.go:51]: Yes we will right after this PR

File: soci/index_builder.go [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+	defaultSpanSize := int64(1 << 22)
+	return &IndexBuilder{
+		contentStore:        contentStore,
+		spanSize:            defaultSpanSize,
+		minLayerSize:        0,
+		buildToolIdentifier: "AWS SOCI CLI v0.1",
+	}
+}
+
+func (b *IndexBuilder) WithSpanSize(spanSize int64) *IndexBuilder {
```
[the remark this answers is not in the document]
Comment (dvnguyen-amzn) [reply soci/index_builder.go:34]: thanks for the link. I'll use that pattern in the next revision

File: soci/index_builder.go (hunk 1 above)
Comment (dvnguyen-amzn) [reply soci/index_builder.go:18]: Agree that it's a big interface. Would you mind if I do a fast follow on that? The reasons are:
1. I don't know how yet
2. I don't want to introduce too much changes in a single PR.

File: soci/index_builder.go (hunk 2 above)
Comment (dvnguyen-amzn) [reply soci/index_builder.go:25]: yes will do in the next revision

File: soci/index_builder.go (hunk 3 above)
Comment (dvnguyen-amzn) [reply soci/index_builder.go:30]: yes will do in the next revision

File: soci/index_builder.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (dvnguyen-amzn) [reply soci/index_builder.go:51]: cc @qingshenhuang

File: soci/index_builder.go [hunk 6]
```diff
@@ -0,0 +1,193 @@
...
+)
+
+type IndexBuilder struct {
+	contentStore        content.Store
+	spanSize            int64
+	minLayerSize        int64
+	buildToolIdentifier string
+}
+
+type IndexBuilderOption func(*IndexBuilder)
```
Comment (rdpsin) [inline soci/index_builder.go:40]: Maybe `IndexBuilderOption` should return an `error`, if an invalid value is ever supplied?

Then we can do


```
for _, opt := range opts {
    if err := opt(b); err != nil {
       return err
     }
}
```

Comment (dvnguyen-amzn) [reply soci/index_builder.go]: done

File: soci/index_builder.go (hunk 1 above)
Comment (rdpsin) [reply soci/index_builder.go:34]: Sure.

File: soci/index_builder.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (rdpsin) [reply soci/index_builder.go:88]: Why not return `*Index` here, instead of `Index`?

File: soci/index_builder.go (hunk 2 above)
Comment (dvnguyen-amzn) [reply soci/index_builder.go]: done

**Review by wmesard — APPROVED** [review]

File: soci/index_builder.go [hunk 7]
```diff
@@ -0,0 +1,198 @@
...
+	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
+	"github.com/pkg/errors"
+	"golang.org/x/net/context"
+	"golang.org/x/sync/errgroup"
+)
+
+const (
+	DefaultSpanSize            = int64(1 << 22)
+	MinLayerSize               = 0
+	DefaultBuildToolIdentifier = "AWS SOCI CLI v0.1"
```
Comment (rdpsin) [inline soci/index_builder.go:36]: nit: Can probably set these to lowercase as they are not be exported from this package.

**Review by rdpsin — APPROVED** [review]

File: soci/index_builder.go (hunk 7 above)
Comment (dvnguyen-amzn) [reply soci/index_builder.go:36]: makes sense. I can do it in the PR

**Review by hanyuel — APPROVED** [review]

**Pull request merged**